pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -704,7 +709,76 @@ func (c *Controller) InstancesByPort(svc *model.Service, reqSvcPort int,
 		return inScopeInstances, nil
 	}
 
-	return c.endpoints.InstancesByPort(c, svc, reqSvcPort, labelsList)
+	outInstances, err := c.endpoints.InstancesByPort(c, svc, reqSvcPort, labelsList)
+	foreignInstances, err2 := c.getForeignServiceInstancesByPort(svc, reqSvcPort)
+
+	if err2 != nil {
+		err = multierror.Append(err, err2)
+	}
+
+	return append(outInstances, foreignInstances...), err
+}
+
+func (c *Controller) getForeignServiceInstancesByPort(svc *model.Service, reqSvcPort int) ([]*model.ServiceInstance, error) {
+	// Run through all the foreign instances, select ones that match these labels
+	// only if this is a kubernetes internal service and of ClientSideLB (eds) type
+	// as InstancesByPort is called by the aggregate controller. We dont want to include
+	// foreign instances for any other registry
+	if svc.Attributes.ServiceRegistry != string(serviceregistry.Kubernetes) ||
+		svc.MeshExternal || svc.Resolution != model.ClientSideLB {
+		return nil, nil
+	}
+
+	item, exists, err := c.services.GetStore().GetByKey(kube.KeyFunc(svc.Attributes.Name, svc.Attributes.Namespace))
+	if err != nil {
+		log.Infof("get foreignServiceInstancesByPort(%s, %s) => error %v", svc.Attributes.Name, svc.Attributes.Namespace, err)
+		return nil, err
+	}
+
+	if !exists {
+		return nil, nil
+	}
+
+	k8sService := item.(*v1.Service)
+	if k8sService.Spec.Selector == nil {
+		// this shouldn't be possible as we already check for resolution above. But just in case
```

Comment:
This is possible,  we can create a service without selectors. And create an endpoint manually for it. This way, it can work like external name service to access remote address.